### PR TITLE
Update upgrade guide 1.4

### PR DESCRIPTION
**Pull request by omarismail**
This updates the 1.4 upgrade guide.

### CODE REVIEW

**Review by apparentlymart — COMMENTED** [review]
With one obvious exception I focused just on the content here since I see you also asked for technical writing review.

As mentioned inline, I'll prepare a PR to update the CLI Configuration page to describe the environment variable form of the `plugin_cache_may_break_dependency_lock_file` setting, so that you can link to there now and it'll all be aligned by the time we publish these docs on the live site.

File: website/docs/language/upgrade-guides/index.mdx
Comment (omarismail) [reply website/docs/language/upgrade-guides/index.mdx:15]: Makes sense, will re-order and remove the last one

File: website/docs/language/upgrade-guides/index.mdx
Comment (omarismail) [reply website/docs/language/upgrade-guides/index.mdx]: Updated

File: website/docs/language/upgrade-guides/index.mdx
Comment (omarismail) [reply website/docs/language/upgrade-guides/index.mdx]: Updated

**Review by apparentlymart — APPROVED** [review]
The content looks good to me now! :tada: 

I'm not going to merge this yet because I only reviewed the content for technical correctness and didn't review for writing style and such; I assume @judithpatudith will approve after she's happy with that, and then I'll merge this.

(I tried to tell Vercel that it should build a preview from this PR but for some reason it accepts my request to do it but then doesn't actually do it. I'm not sure what's up with that :man_shrugging: ... we'll probably want to find _some_ way to see a rendered preview of this before we move forward to make sure there aren't any formatting or linking quirks specific to the `developer.hashicorp.com` platform, which has some slightly different Markdown expectations than GitHub does. Hopefully the digital marketing web team can poke at this and make the preview render, since they have more privileged access on Vercel than I do.)

File: website/docs/language/upgrade-guides/index.mdx
Comment (judithpatudith) [inline website/docs/language/upgrade-guides/index.mdx:8]:
```suggestion
-> **Tip:** Use the version selector to view the upgrade guides for older Terraform versions.
```

File: website/docs/language/upgrade-guides/index.mdx
Comment (judithpatudith) [inline website/docs/language/upgrade-guides/index.mdx:12]:
```suggestion
Terraform v1.4 honors the [Terraform v1.0 Compatibility Promises](https://developer.hashicorp.com/terraform/language/v1-compatibility-promises), but there are some behavior changes outside of those promises that may affect a small number of users. Specifically, the following updates may require additional upgrade steps:
```

File: website/docs/language/upgrade-guides/index.mdx
Comment (judithpatudith) [inline website/docs/language/upgrade-guides/index.mdx:20]:
```suggestion
`terraform init` now ignores entries in the optional global provider cache directory unless they match a checksum already tracked in the current configuration's dependency lock file. 
```

File: website/docs/language/upgrade-guides/index.mdx
Comment (judithpatudith) [inline website/docs/language/upgrade-guides/index.mdx:24]:
```suggestion
For more details and how to keep using the prior incorrect behavior, please see the [documentation](/cli/config/config-file#allowing-the-provider-plugin-cache-to-break-the-dependency-lock-file).
```

File: website/docs/language/upgrade-guides/index.mdx
Comment (judithpatudith) [inline website/docs/language/upgrade-guides/index.mdx:28]:
```suggestion
When interpreting the hostname portion of a provider source address or the address of a module in a module registry, Terraform now uses non-transitional IDNA2008 mapping rules instead of the transitional mapping rules. Terraform no longer accepts the characters `ß` (U+00DF, "LATIN SMALL LETTER SHARP S") and `ς` (U+03C2, "GREEK SMALL LETTER FINAL SIGMA") Use the available alternative forms for both characters instead.
```

Can we say what the alternative forms are, or is it reasonable to expect that readers already know them?

File: website/docs/language/upgrade-guides/index.mdx
Comment (judithpatudith) [inline website/docs/language/upgrade-guides/index.mdx:30]:
```suggestion
This change better adheres to the [the WHATWG URL spec's rules for interpreting non-ASCII domain names](https://url.spec.whatwg.org/#concept-domain-to-ascii). Terraform tries to interpret host names the same way that web browsers do. For some hostnames containing non-ASCII characters this may cause Terraform to now request a different "punycode" hostname when resolving. 
```

**Review by judithpatudith — APPROVED** [review]
Thanks for the opportunity to review! I added some suggestions, but there was one spot where I wasn't sure what to suggest so I wrote a madlib instead! Approved!

File: website/docs/language/upgrade-guides/index.mdx
Comment (apparentlymart) [reply website/docs/language/upgrade-guides/index.mdx:28]: My interpretation of the specification was that these two letters are something like "archaic forms" that are not typically used in modern text written in the languages they come from (I assume German for the sharp S and Greek for the Greek letter), and so I was guessing that folks who speak those languages would know about this already, but that is a big assumption since I have incomplete knowledge of this myself.

Unfortunately here we're mostly just passing on a change made in an upstream specification so I was just doing my best to translate the information in that spec to be relevant to Terraform. :confounded:

File: website/docs/language/upgrade-guides/index.mdx
Comment (omarismail) [reply website/docs/language/upgrade-guides/index.mdx:22]: this is correct, ty!

**Comment by apparentlymart** [comment]
It looks like this is in a good state for initial release so I'm going to merge it now. We often tweak the upgrade guide after initial release once we learn what sorts of questions folks have about upgrading, so I'm sure we'll get an opportunity to revise this more if we need to.

**Comment by crw** [comment]
Should we say more about the `plugin_cache_may_break_dependency_lock_file`? This seemed to be a fairly contentious change amongst the people who are impacted, it might be worth just clarifying the situation in which someone would be impacted by this change and would need to update their usage patterns or choose to use the workaround.

**Comment by apparentlymart** [comment]
Oh, sorry @crw your comment appeared only after I pressed the button. :roll_eyes: 

If the answer to that question turns out to be "yes", let's revise that in a separate PR.

**Pull request merged**
